Initial implementation

rafecolton opened the pull request:
Do not merge until:
- [x] at least 2 :shipit:

Commit:
Adding ref to genome documentation

Commit:
Changing "ret" to "subject" in genome test

Commit:
Filling in cloning functionality and tests

Commit:
DRYing things up a little

Commit:
[OOPS] mot passing --depth if depth not specified

jszwedko commented:
I'm curious to see what the intent of this library is, but one thing that jumped out at me was the use of a map for options which feels a unidiomatic to me, especially since some fields are required. You may consider using named arguments for the required fields and something like what Rob Pike describes in http://commandcenter.blogspot.nl/2014/01/self-referential-functions-and-design.html?m=1 for the optional ones. Just a thought though, I haven't found what feels like a great pattern for this yet; just that a map feels a little messy.

jszwedko commented:
Never mind, I see the intent from the diff :) was only looking at what is in master.

rafecolton commented:
@jszwedko I agree with you, this does feel a bit messy.  In particular, having the `genome` type be private felt odd as did having the validation inside of `NewGenome()` (or having a `NewGenome()` function at all).  I think the current organization is actually a function of me writing that part of the code first.  I think this can be dramatically simplified by moving the validation inside of the `Clone()` function.  Let me see what I can whip up.

Also, the intent for this library is to be used in the docker build server :smile:

Commit:
Dramatically simplifying Genome / option passing

rafecolton commented:
@jszwedko what do you think about the latest commit?  Should simplify things a bit as well as making them more idiomatic (plus I'm a lot happier about it than using a `map`)

jszwedko commented:
@rafecolton Figured this was for the docker build server! Definitely feels much better to me after your last commit.

rafecolton commented:
@jszwedko :thumbsup:

I'd love to get your thoughts on [this](https://github.com/modcloth/kamino/blob/initial-implementation/clone_creator.go#L112-L175) function.  It's come up in a couple other contexts as well and I have yet to find the magical incantation.

Commit:
Removing some unnecessary lines

Commit:
Being a little smarter about detecting a detached HEAD in the cache dir

Commit:
Making our git operations more `--quiet`

Commit:
Oops, making comment more accurate [ci skip]

rafecolton commented:
@jszwedko @meatballhat @jtwb would you guys mind taking a look and sanity checking this?  In particular, I'm looking for feedback on the [`updateToRef()`](https://github.com/modcloth/kamino/blob/initial-implementation/clone_creator.go#L112-L175) function in `clone_creator.go`.

radkin commented:
:thumbsup:

meatballhat commented on clone_creator.go:
```diff
@@ -0,0 +1,175 @@
+		return err
+	}
+
+	git, err := fileutils.Which("git")
```
You're checking for a `git` in `$PATH` here, but not above.  Is that right?

rafecolton replied:
Yep - `exec.Command()` will find the executable in `$PATH` for you and populate it as  `Path:` in the `exec.Cmd` that gets returned.  If you are creating an `&exec.Cmd{}` literal, you have to find it yourself.

rafecolton replied:
Also worth noting that I'm using `exec.Comand()` above because it's a little easier if there are no special options and an `exec.Cmd` literal below to specify `Dir:`

meatballhat commented on clone_factory.go:
```diff
@@ -0,0 +1,86 @@
+)
+
+/*
+A CloneFactory generates your clones for you.  Create a clone factory with NewCloneFactory().
```
beware the F word :trollface:

Are you sure this shouldn't be a `CloneManagerProxyController`?

rafecolton replied:
Haha fair point.  I was thinking a literal factory in the spirit of the namesake.  Do you think `CloneFacility` would still be too widget-y?

meatballhat commented on clone_factory.go:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	creator := &clone{
+		Genome:  g,
```
You're naming the `Genome:` member here, but embedding the `*Genome` member in the `clone` struct.  Are you using any of the `*Genome` methods?

rafecolton replied:
Not methods, but I'm using the `*Genome` fields all throughout `clone_creator.go`.  In the first iteration, I was doing things like `clone.g.Repo` and that got old real quick

meatballhat replied:
so it's a convenience thing?  :trollface:

rafecolton replied:
Yes indeed!

Also, conceptually it makes sense - one would say, "That person has blue eyes," and not, "That person's genome has blue eyes."  :trollface:

jszwedko commented on clone_creator.go:
```diff
@@ -0,0 +1,175 @@
+		},
+	}
+
+	for _, cmd := range cmds {
```
Seems reasonable to me (though I think `git checkout --force` will handle the `git reset` for you unless there is an edge case I'm missing.

rafecolton replied:
Yep, good call - it looks like the two would have the same effect since I'm just using `git reset --hard` to throw away any local changes and not doing any merges or making any commits.

Commit:
Removing `git reset --hard`, since it is redundant

It is only being used to throw away any local changes, and since we are
not doing any commits on the detached head, `git checkout --force` has
the same effect
